### PR TITLE
Bpf: Add missing policy match cases

### PR DESCRIPTION
**Pull request by jrajahalme**
Socket level egress policy check is missing the L4 protocol match to any
destination, while the packet level policy check is missing the case
where protocol is matched without a port with a specific destination or
source, for egress or ingress, respectively. Add those missing cases.

Packet level match on L4 protocol with wildcarded port number can also
be done on untracked fragments, as the missing port number is not needed
in this case.

Add policy match types for protocol matches with wildcarded port so that
the different match cases are not conflated in flow records.

Fixes: #21185
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
Fix issues with policy handling introduced by new policy match support for L4 ports on any protocol.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
/test-1.16-4.9

File: bpf/lib/policy.h [hunk 1]
```diff
@@ -30,62 +30,65 @@ policy_sk_egress(__u32 identity, __u32 ip,  __u16 dport)
 	if (!map)
 		return CTX_ACT_OK;
 
+	/* Policy match precedence:
```
Comment (ldelossa) [inline bpf/lib/policy.h:33]: Maybe worth ditching the first commit all together, and keeping this one with the updated comments?

**Review by ldelossa — APPROVED** [review]
Changes LGTM, just a nit on first commit basically being obsoleted by following.

File: bpf/lib/policy.h [hunk 2]
```diff
@@ -71,7 +71,19 @@ policy_sk_egress(__u32 identity, __u32 ip,  __u16 dport)
...
+	/* Check Proto-only policy */
+	key.sec_label = 0;
+	policy = map_lookup_elem(map, &key);
+	if (likely(policy)) {					/* 4. ANY/proto/ANY */
+		/* FIXME: Need byte counter */
+		__sync_fetch_and_add(&policy->packets, 1);
+		if (unlikely(policy->deny))
+			return DROP_POLICY_DENY;
+		return CTX_ACT_OK;
+	}
```
Comment (joestringer) [inline bpf/lib/policy.h:67]: It'd be nice to avoid these extra lookups if we don't have such policies or something. Even on the original PR that introduced this support we had a bit of discussion around the performance impact, and now we're adding yet more checks to the policy code for every packet. That said, I don't think such a change is appropriate for this PR. I'll defer to @cilium/sig-datapath on how much they want to pursue performance optimization on this front.

**Review by joestringer — APPROVED** [review]
Thanks for the fix, LGTM.

**Comment by joestringer** [comment]
Incidentally, this PR suggests that we're also lacking test cases for this scenario since this is only making datapath changes. Perhaps that's also an item for @cilium/sig-datapath to follow up on?

**Review by pchaigno — CHANGES_REQUESTED** [review]
Changes look good to me (assuming verifier test is still passing with the increased complexity).

We however need release notes for the three bugs fixed here (fragment, socket-level policy, bpf_lxc policy).

**Comment by pchaigno** [comment]
Also, shouldn't this be `release-note/bug` and backported to v1.12?

File: bpf/lib/policy.h [hunk 3]
```diff
@@ -174,16 +182,23 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
...
+	 * 6. ANY/ANY/ANY
+	 */
 	/* L4 lookup can't be done on untracked fragments. */
 	if (!is_untracked_fragment) {
 		/* Start with L3/L4 lookup. */
 		policy = map_lookup_elem(map, &key);
 		if (likely(policy)) {
 			cilium_dbg3(ctx, DBG_L4_CREATE, remote_id, local_id,
 				    dport << 16 | proto);
-
```
Comment (sahid) [inline bpf/lib/policy.h:184]: nit: unrelated change.

File: bpf/lib/policy.h (hunk 2 above)
Comment (sahid) [reply bpf/lib/policy.h:67]: Should not we create an issue for that? At least to keep track of your point and perhaps if some contributors have interest, open the discussion...

File: bpf/lib/policy.h [hunk 4]
```diff
@@ -41,71 +41,54 @@ policy_sk_egress(__u32 identity, __u32 ip,  __u16 dport)
...
-		__sync_fetch_and_add(&policy->packets, 1);
-		if (unlikely(policy->deny))
-			return DROP_POLICY_DENY;
-		return CTX_ACT_OK;
+		goto policy_check_entry;
 	}
 
 	return DROP_POLICY;
+
+policy_check_entry:
```
Comment (sahid) [inline bpf/lib/policy.h:86]: Looks like a reasonable improvement even if that sounds touchy considering it's a release blocker.

**Review by sahid — APPROVED** [review]

File: bpf/lib/policy.h (hunk 2 above)
Comment (joestringer) [reply bpf/lib/policy.h:67]: Looking back, while I raised this concern previously, I didn't see any interest raised from sig-datapath folks. I don't know for sure what the impact of this particular set of changes is vs. the rest of the datapath, so maybe it's a bit premature to raise an issue specifically for this. It probably makes more sense that we look at performance more holistically and just assess performance in general. As part of that, we can follow up wherever the low-hanging fruit are. We can certainly file an issue for "investigating network performance bottlenecks in Cilium", but unless someone is actively interested and looking into it, then I'm not sure whether it matters to raise the issue now. When someone's interested they can file an issue and we can discuss the findings.

**Comment by pchaigno** [comment]
> > Also, shouldn't this be release-note/bug and backported to v1.12?
> 
> AFAIK this is only fixing #21185 which was introduced in v1.13.0-rc1.

Oh, right! And all three bugs where introduced by that PR. Never mind me then :slightly_smiling_face:

File: bpf/lib/policy.h (hunk 4 above)
Comment (joestringer) [reply bpf/lib/policy.h:86]: This now matches the packet egress case, and the change in general is fixing a bug that's new to v1.13.0 RCs. I'm OK with a little reordering here.
